Fix RecursionError in Husqvarna Automower coordinator

Thomas55555 opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Instead of recursing in the same task the coordinator creates a new background task for the call to `client_listen` on retry

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #123008
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

Thomas55555 committed:
reach maximum recursion depth exceeded in tests

Thomas55555 committed:
second background task

MartinHjelmare commented on homeassistant/components/husqvarna_automower/coordinator.py:
```diff
@@ -69,19 +69,18 @@ async def client_listen(
-            )
+            if not hass.is_stopping:
+                reconnect_time = 2
+                await asyncio.sleep(reconnect_time)
```
We may need to store the `reconnect_time` in an instance variable so we can reset it after a successful connection. Although, we will never retry now if the connection is successful even if the listening call fails. We should probably retry again in that case too. Otherwise the websocket connection will never recover until the config entry is reloaded.

MartinHjelmare commented on tests/components/husqvarna_automower/test_init.py:
```diff
@@ -128,20 +131,29 @@ async def test_websocket_not_available(
+    reconnect_time = 2
+    for count in range(1, 945):
+        reconnect_time = min(reconnect_time * 2, MAX_WS_RECONNECT_TIME)
+        freezer.tick(timedelta(seconds=reconnect_time))
```
We can't test it like this since we don't want to sleep for any time in tests. We could make the default reconnect time a constant in the integration to make it easier to patch, then patch that to zero with the `new=0` parameter, and then test that we don't hit the recursion limit.

Thomas55555 replied:
If we patch the time to 0, from the beginning of this test, then it depends on the machines performance, how many retries we've already have, until the loop starts. So the test is very flakey. Do we actually have to test it?

Thomas55555 replied:
I can just reset the mock..

Thomas55555 committed:
Update homeassistant/components/husqvarna_automower/coordinator.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

Thomas55555 committed:
Update homeassistant/components/husqvarna_automower/coordinator.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

Thomas55555 committed:
test

Thomas55555 committed:
modify test

Thomas55555 committed:
tests

Thomas55555 committed:
use correct exception

Thomas55555 commented on homeassistant/components/husqvarna_automower/coordinator.py:
```diff
@@ -64,24 +67,22 @@ async def client_listen(
-        except HusqvarnaWSServerHandshakeError as err:
+            await automower_client.start_listening()  # Ensure we catch errors here too
+            self.reconnect_time = DEFAULT_RECONNECT_TIME  # Reset reconnect time after successful connection
+        except (HusqvarnaWSServerHandshakeError, Exception) as err:
```
We don't expect that. I just haven't used the right Exception. It's changed now.

Thomas55555 committed:
reset mock

Thomas55555 committed:
use recursion_limit

Thomas55555 committed:
remove unneeded ticks

Thomas55555 committed:
test TimeoutException

MartinHjelmare commented on tests/components/husqvarna_automower/test_init.py:
```diff
@@ -127,21 +128,44 @@ async def test_websocket_not_available(
+
+        # Initial call count and range for reconnection attempts
+        mock_automower_client.reset_mock()
+        recursion_limit = sys.getrecursionlimit()
```
I'd set the recursion limit temporarily to a low number for this test via `sys.setrecursionlimit`.

https://docs.python.org/3/library/sys.html#sys.setrecursionlimit

MartinHjelmare commented on homeassistant/components/husqvarna_automower/coordinator.py:
```diff
@@ -64,24 +68,27 @@ async def client_listen(
-            reconnect_time = 2
-            await automower_client.start_listening()
+            await automower_client.start_listening()  # Ensure we catch errors here too
+            self.reconnect_time = DEFAULT_RECONNECT_TIME  # Reset reconnect time after successful connection
```
Move the comment above the code line so the lines are kept within 88 characters.

Thomas55555 replied:
done in https://github.com/home-assistant/core/pull/123085/commits/40632e3aea0b48caab264c9c461f03f1994cfffc

MartinHjelmare commented on tests/components/husqvarna_automower/test_init.py:
```diff
@@ -127,21 +128,44 @@ async def test_websocket_not_available(
+        assert (
+            mock_automower_client.auth.websocket_connect.call_count
+            == recursion_limit + 2
+        )
```
I'd reset the side effect again here and test that we connect and listen successfully.

MartinHjelmare replied:
And maybe finish the test by having another exception and then stopping `hass` with `hass.async_stop` and checking that we don't try to connect again.

Thomas55555 replied:
> I'd reset the side effect again here and test that we connect and listen successfully.

done in https://github.com/home-assistant/core/pull/123085/commits/196bbb684ea1633ee527406decd2e2d43a602f06
We test the successful connection by checking the caplog. Is that okay?

Thomas55555 replied:
> And maybe finish the test by having another exception and then stopping `hass` with `hass.async_stop` and checking that we don't try to connect again.

Done in https://github.com/home-assistant/core/pull/123085/commits/84378003f1c77a5afde77363f4547ee657690f2c

Thomas55555 committed:
set lower recursionlimit

Thomas55555 committed:
remove not that important comment and move the other

Thomas55555 committed:
test that we connect and listen successfully

Thomas55555 committed:
Simulate hass shutting down

MartinHjelmare commented on tests/components/husqvarna_automower/test_init.py:
```diff
@@ -127,21 +128,74 @@ async def test_websocket_not_available(
+        # Initial call count and range for reconnection attempts
+        mock_automower_client.reset_mock()
+        default_recursion_limit = sys.getrecursionlimit()
+        temporary_recursion_limit = 200
```
I was thinking in the order of magnitude of 10.

Or we just skip testing against the recursion limit. It's not like we always test against the recursion limit when we have something that repeats.

Thomas55555 replied:
On my machine we already have 44 recursions while setting up the integration. I think it's better to skip the tests, although it was a lot of work.

Thomas55555 committed:
skip testing against the recursion limit

MartinHjelmare commented on tests/components/husqvarna_automower/test_init.py:
```diff
@@ -127,21 +127,50 @@ async def test_websocket_not_available(
+        new=0,
+    ):
+        # Simulate a WebSocket handshake error
+        mock_automower_client.auth.websocket_connect.side_effect = (
```
I think we should customize the side effect with a function so that the first task raises the exception but then in the second task we block, eg by waiting for an `asyncio.Event` or `asyncio.Future`. Otherwise there'll be a fast loop creating tasks here.

Thomas55555 replied:
Sorry, I don't understand what you mean. Can you try to explain that with different words. Or would it also possible to not patch the `DEFAULT_RECONNECT_TIME` not to zero, to avoid the changes?

MartinHjelmare replied:
Here's an example of how to set a side effect that blocks until an event (`listen_block`) is set by the test.

https://github.com/home-assistant/core/blob/866ffcf639d3be0f4e6a35884b41fa3e6bff83b2/tests/components/zwave_js/conftest.py#L500-L526

Thomas55555 committed:
Merge branch 'dev' into max-recursion

Thomas55555 committed:
Update homeassistant/components/husqvarna_automower/coordinator.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

Thomas55555 committed:
Merge branch 'dev' into max-recursion

Thomas55555 committed:
mock

emontnemery commented:
@Thomas55555 You need to click the "Ready for review"-button if you're done with addressing review comments and you want the PR to be reviewed again.

Thomas55555 committed:
Remove comment

Thomas55555 commented:
Thanks, I know. But I haven't addressed all comments. And I'm not sure I can.
But I also don't want to close the PR, because the issue is fixed and only the tests are not right.

emontnemery commented:
> Thanks, I know. But I haven't addressed all comments. And I'm not sure I can. But I also don't want to close the PR, because the issue is fixed and only the tests are not right.

Oh, I see.
If you need some help, feel free to contact me on Discord, my nick there is @emontnemery

emontnemery committed:
Revert "mock"

This reverts commit e8ddaea3d79ed1aceb696a055cc42ad08b4febca.

emontnemery committed:
Move patch to decorator

emontnemery committed:
Make execution of patched methods predictable

emontnemery committed:
Merge remote-tracking branch 'upstream/dev' into max-recursion

MartinHjelmare commented on tests/components/husqvarna_automower/test_init.py:
```diff
@@ -127,28 +129,83 @@ async def test_update_failed(
+
+    # Simulate a successful connection
+    caplog.clear()
+    await mock_called.wait()
```
It looks like the connect call is now blocking on the mock_stall event again after this await.

MartinHjelmare commented on tests/components/husqvarna_automower/test_init.py:
```diff
@@ -127,28 +129,83 @@ async def test_update_failed(
+    mock_automower_client.auth.websocket_connect.side_effect = None
+    mock_automower_client.start_listening.side_effect = mock_function
+    # Check outcome
     await hass.async_block_till_done()
```
So this await doesn't seem needed.

MartinHjelmare commented on tests/components/husqvarna_automower/test_init.py:
```diff
@@ -127,28 +129,83 @@ async def test_update_failed(
+    mock_automower_client.start_listening.side_effect = mock_function
+    # Check outcome
     await hass.async_block_till_done()
     assert mock_automower_client.auth.websocket_connect.call_count == 2
+    assert mock_automower_client.start_listening.call_count == 0
+    assert "Trying to reconnect: Boom" not in caplog.text
```
These asserts should move up before changing the side effects I think to make it more clear.

MartinHjelmare commented on tests/components/husqvarna_automower/test_init.py:
```diff
@@ -127,28 +129,83 @@ async def test_update_failed(
     assert mock_automower_client.auth.websocket_connect.call_count == 2
+    assert mock_automower_client.start_listening.call_count == 0
+    assert "Trying to reconnect: Boom" not in caplog.text
+    mock_stall.set()
```
Setting the mock_stall event here will raise the same exception for the websocket_connect call as we tested above, even though we have now changed the side effects of the patched functions, since this change was done after.

MartinHjelmare replied:
`mock_stall` is never cleared.
```suggestion
```

emontnemery committed:
Parametrize test, make mocked start_listening block

MartinHjelmare commented:
Did you intend to push some commits?

emontnemery commented:
> Did you intend to push some commits?

Yes, commit 564f4d1 👍

MartinHjelmare commented on homeassistant/components/husqvarna_automower/coordinator.py:
```diff
@@ -66,24 +68,28 @@ async def client_listen(
         try:
             await automower_client.auth.websocket_connect()
-            reconnect_time = 2
             await automower_client.start_listening()
+            # Reset reconnect time after successful connection
+            self.reconnect_time = DEFAULT_RECONNECT_TIME
```
We need to reset the reconnect time before the listen call since that will block if it's successful.
```suggestion
            # Reset reconnect time after successful connection
            self.reconnect_time = DEFAULT_RECONNECT_TIME
            await automower_client.start_listening()
```

MartinHjelmare committed:
Apply suggestions from code review

MartinHjelmare commented on tests/components/husqvarna_automower/test_init.py:
```diff
@@ -127,28 +129,77 @@ async def test_update_failed(
+        ),
+        (
+            ["start_listening"],
+            TimeoutException,
```
Side note: I'm not sure that the exceptions set here are specific to each of the methods. Both exception could maybe happen for both methods.

We can address that in a follow up, since it's just a test improvement.

MartinHjelmare reviewed (approved):
Thanks!
